File: static/offline.js
/**
 * Offline Support JavaScript
 * Registers service worker and handles offline functionality
 */

(function() {
  'use strict';
  
  // Check if service workers are supported
  if ('serviceWorker' in navigator) {
    // Register service worker
    window.addEventListener('load', () => {
      navigator.serviceWorker.register('/static/service-worker.js')
        .then((registration) => {
          console.log('[PWA] Service Worker registered:', registration.scope);
          
          // Check for updates
          registration.addEventListener('updatefound', () => {
            const newWorker = registration.installing;
            newWorker.addEventListener('statechange', () => {
              if (newWorker.state === 'installed' && navigator.serviceWorker.controller) {
                // New service worker available
                console.log('[PWA] New service worker available');
                showUpdateNotification();
              }
            });
          });
        })
        .catch((error) => {
          console.error('[PWA] Service Worker registration failed:', error);
        });
      
      // Listen for controller changes
      navigator.serviceWorker.addEventListener('controllerchange', () => {
        console.log('[PWA] Service Worker controller changed');
        window.location.reload();
      });
    });
  }
  
  // Online/Offline status indicator
  function createOfflineIndicator() {
    // Check if indicator already exists
    if (document.getElementById('pwa-offline-indicator')) {
      return;
    }
    
    const indicator = document.createElement('div');
    indicator.id = 'pwa-offline-indicator';
    indicator.style.cssText = `
      position: fixed;
      top: 0;
      left: 0;
      right: 0;
      background: #f44336;
      color: white;
      text-align: center;
      padding: 10px;
      z-index: 10000;
      display: none;
      font-weight: 500;
      box-shadow: 0 2px 4px rgba(0,0,0,0.2);
    `;
    indicator.textContent = '⚠️ Bạn đang offline - Một số tính năng có thể không khả dụng';
    document.body.insertBefore(indicator, document.body.firstChild);
    
    // Update indicator based on online status
    function updateIndicator() {
      if (!navigator.onLine) {
        indicator.style.display = 'block';
      } else {
        indicator.style.display = 'none';
      }
    }
    
    window.addEventListener('online', updateIndicator);
    window.addEventListener('offline', updateIndicator);
    updateIndicator();
  }
  
  // Install prompt for PWA
  let deferredPrompt;
  
  window.addEventListener('beforeinstallprompt', (e) => {
    // Prevent the mini-infobar from appearing
    e.preventDefault();
    // Stash the event so it can be triggered later
    deferredPrompt = e;
    // Show install button (you can customize this)
    showInstallPrompt();
  });
  
  function showInstallPrompt() {
    // Create install button if it doesn't exist
    if (document.getElementById('pwa-install-button')) {
      return;
    }
    
    const installBtn = document.createElement('button');
    installBtn.id = 'pwa-install-button';
    installBtn.textContent = '📱 Cài đặt ứng dụng';
    installBtn.style.cssText = `
      position: fixed;
      bottom: 20px;
      right: 20px;
      background: #1976d2;
      color: white;
      border: none;
      padding: 12px 20px;
      border-radius: 25px;
      cursor: pointer;
      font-size: 14px;
      font-weight: 500;
      box-shadow: 0 4px 12px rgba(0,0,0,0.3);
      z-index: 9999;
      display: none;
    `;
    
    installBtn.addEventListener('click', async () => {
      if (deferredPrompt) {
        // Show the install prompt
        deferredPrompt.prompt();
        // Wait for the user to respond
        const { outcome } = await deferredPrompt.userChoice;
        console.log('[PWA] User choice:', outcome);
        // Clear the deferred prompt
        deferredPrompt = null;
        installBtn.style.display = 'none';
      }
    });
    
    document.body.appendChild(installBtn);
    
    // Show button after a delay
    setTimeout(() => {
      if (deferredPrompt) {
        installBtn.style.display = 'block';
      }
    }, 3000);
  }
  
  // Initialize when DOM is ready
  if (document.readyState === 'loading') {
    document.addEventListener('DOMContentLoaded', () => {
      createOfflineIndicator();
    });
  } else {
    createOfflineIndicator();
  }
  
  // Cache API responses (optional - for drug database, etc.)
  function cacheAPIResponse(url, data) {
    if ('caches' in window) {
      caches.open('clinical-assistant-v1').then((cache) => {
        const response = new Response(JSON.stringify(data), {
          headers: { 'Content-Type': 'application/json' }
        });
        cache.put(url, response);
      });
    }
  }
  
  // Expose cache function globally (for use in Streamlit)
  window.cacheAPIResponse = cacheAPIResponse;
  
  // Check if app is installed
  function isInstalled() {
    return window.matchMedia('(display-mode: standalone)').matches ||
           window.navigator.standalone === true;
  }
  
  // Expose utility functions
  window.PWA = {
    isInstalled: isInstalled,
    cacheAPIResponse: cacheAPIResponse,
    updateServiceWorker: () => {
      if ('serviceWorker' in navigator && navigator.serviceWorker.controller) {
        navigator.serviceWorker.controller.postMessage({ type: 'SKIP_WAITING' });
      }
    }
  };
  
  console.log('[PWA] Offline support initialized');
})();

File: components/offline.py
"""
Offline Support Component
Displays offline status indicator and PWA install prompt
"""

import streamlit as st
from pathlib import Path


def render_offline_indicator():
    """
    Render enhanced offline status indicator in the UI
    Shows when user is offline with better styling and information
    """
    st.markdown(
        """
        <div id="streamlit-offline-indicator" style="display: none;">
            <div style="
                background: linear-gradient(135deg, #f44336 0%, #d32f2f 100%); 
                color: white; 
                padding: 12px 20px; 
                text-align: center; 
                position: fixed; 
                top: 0; 
                left: 0; 
                right: 0; 
                z-index: 10000; 
                box-shadow: 0 4px 6px rgba(0,0,0,0.3);
                font-weight: 500;
                display: flex;
                align-items: center;
                justify-content: center;
                gap: 10px;
            ">
                <span style="font-size: 1.2em;">⚠️</span>
                <span>Bạn đang offline - Drug Database đã cache vẫn hoạt động</span>
            </div>
        </div>
        <style>
            /* Add padding to body when offline indicator is shown */
            body:has(#streamlit-offline-indicator[style*="display: block"]) {
                padding-top: 50px;
            }
        </style>
        <script>
            // Update indicator based on online status
            function updateOfflineIndicator() {
                const indicator = document.getElementById('streamlit-offline-indicator');
                if (indicator) {
                    if (!navigator.onLine) {
                        indicator.style.display = 'block';
                        // Add padding to main content
                        document.body.style.paddingTop = '50px';
                    } else {
                        indicator.style.display = 'none';
                        document.body.style.paddingTop = '0';
                    }
                }
            }
            
            window.addEventListener('online', () => {
                updateOfflineIndicator();
                // Show brief success message
                console.log('[PWA] Back online');
            });
            window.addEventListener('offline', () => {
                updateOfflineIndicator();
                console.log('[PWA] Gone offline');
            });
            updateOfflineIndicator();
        </script>
        """,
        unsafe_allow_html=True
    )


def render_pwa_info():
    """
    Render PWA installation info and status
    """
    with st.expander("📱 Cài Đặt Ứng Dụng (PWA)", expanded=False):
        st.markdown("""
        **Progressive Web App (PWA)** - Cài đặt ứng dụng để:
        - ✅ Hoạt động offline
        - ✅ Truy cập nhanh từ màn hình chính
        - ✅ Trải nghiệm như app native
        
        **Cách cài đặt:**
        1. **Chrome/Edge:** Click icon "Cài đặt" trong thanh địa chỉ
        2. **Safari (iOS):** Share → Add to Home Screen
        3. **Firefox:** Menu → Install
        
        **Tính năng offline (Enhanced):**
        - ✅ Cache static assets (CSS, JS, images)
        - ✅ Cache drug database đã tải (348+ thuốc)
        - ✅ Offline search trong drug database
        - ✅ Cache calculator definitions
        - ✅ Offline fallback page khi mất kết nối
        - ✅ Service Worker tự động cache dữ liệu
        
        **💡 Mẹo sử dụng offline:**
        - Truy cập Drug Database khi online để cache dữ liệu
        - Search và filters sẽ hoạt động với dữ liệu đã cache
        - Drug detail pages sẽ hiển thị nếu đã tải trước đó
        """)
        
        # Check if PWA is installed
        st.markdown(
            """
            <script>
            function checkPWAInstalled() {
                const isInstalled = window.matchMedia('(display-mode: standalone)').matches ||
                                   window.navigator.standalone === true;
                if (isInstalled) {
                    window.parent.postMessage({type: 'pwa-installed', value: true}, '*');
                }
            }
            checkPWAInstalled();
            </script>
            """,
            unsafe_allow_html=True
        )
        
        # Show cache status
        if st.button("🔄 Xóa Cache & Tải Lại", help="Xóa cache và tải lại service worker"):
            st.markdown(
                """
                <script>
                if ('serviceWorker' in navigator) {
                    navigator.serviceWorker.getRegistrations().then(function(registrations) {
                        for(let registration of registrations) {
                            registration.unregister();
                        }
                        caches.keys().then(function(names) {
                            for (let name of names) {
                                caches.delete(name);
                            }
                        });
                        alert('Cache đã được xóa. Trang sẽ tự động tải lại...');
                        window.location.reload();
                    });
                }
                </script>
                """,
                unsafe_allow_html=True
            )


def render_offline_status():
    """
    Render enhanced current offline/online status with cache information
    """
    st.markdown(
        """
        <div id="pwa-status-indicator" style="padding: 12px; border-radius: 8px; margin: 10px 0; border: 1px solid #e0e0e0;">
            <div style="display: flex; align-items: center; justify-content: space-between; flex-wrap: wrap; gap: 10px;">
                <div>
                    <strong>Trạng thái kết nối:</strong> 
                    <span id="pwa-status-text">Đang kiểm tra...</span>
                </div>
                <div id="pwa-cache-status" style="font-size: 0.9em; color: #666;">
                    <span id="pwa-cache-text">Cache: Đang kiểm tra...</span>
                </div>
            </div>
        </div>
        <script>
            function updateStatus() {
                const statusEl = document.getElementById('pwa-status-indicator');
                const textEl = document.getElementById('pwa-status-text');
                
                if (navigator.onLine) {
                    statusEl.style.background = '#e8f5e9';
                    statusEl.style.borderColor = '#4caf50';
                    textEl.innerHTML = '✅ <strong>Online</strong> - Kết nối internet hoạt động';
                } else {
                    statusEl.style.background = '#fff3cd';
                    statusEl.style.borderColor = '#ffc107';
                    textEl.innerHTML = '⚠️ <strong>Offline</strong> - Drug Database đã cache vẫn hoạt động';
                }
            }
            
            function updateCacheStatus() {
                const cacheTextEl = document.getElementById('pwa-cache-text');
                if (!cacheTextEl) return;
                
                if ('caches' in window) {
                    caches.keys().then(function(names) {
                        const hasCache = names.length > 0;
                        if (hasCache) {
                            cacheTextEl.innerHTML = '💾 Cache: <strong style="color: #4caf50;">Đã có</strong> (' + names.length + ' cache)';
                        } else {
                            cacheTextEl.innerHTML = '💾 Cache: <strong style="color: #ff9800;">Chưa có</strong>';
                        }
                    }).catch(function() {
                        cacheTextEl.innerHTML = '💾 Cache: Không xác định';
                    });
                } else {
                    cacheTextEl.innerHTML = '💾 Cache: Không hỗ trợ';
                }
            }
            
            window.addEventListener('online', updateStatus);
            window.addEventListener('offline', updateStatus);
            updateStatus();
            updateCacheStatus();
            
            // Check service worker
            if ('serviceWorker' in navigator) {
                navigator.serviceWorker.ready.then(function(registration) {
                    if (registration) {
                        const statusText = document.getElementById('pwa-status-text');
                        if (statusText && navigator.onLine) {
                            statusText.innerHTML += '<br><small style="font-weight: normal;">✅ Service Worker đã sẵn sàng - Offline mode enabled</small>';
                        }
                    }
                    updateCacheStatus();
                });
                
                // Also check after a delay
                setTimeout(updateCacheStatus, 2000);
            }
        </script>
        """,
        unsafe_allow_html=True
    )

